minimal_scheduling/README.md

```diff
@@ -0,0 +1,140 @@
+creation. The thread middleware threads will inherit these settings because they are created after
+the scheduling configuration.
+
+This could be interesting when middleware communications shall not participate in the real-time
+execution. Also, this could be useful with some DDS implementations which allow to configure
+each thread scheduling individually, allowing to set a priority different to the main thread.
+
+```bash
+$ ros2 run minimal_scheduling minimal_sched_fifo_middleware_threads
+```
+
+Use `htop` or `ps` to observe the priorities of the threads.
+
+```bash
+$ ps -C minimal_sched_f -L -o tid,comm,rtprio,cls,psr
+    TID COMMAND         RTPRIO CLS PSR
+   6794 minimal_sched_f     80  FF   2
+   6795 minimal_sched_f     80  FF   6
+   6796 gc                  80  FF   6
+   6797 dq.builtins         80  FF   3
+   6798 dq.user             80  FF   6
+   6799 tev                 80  FF   6
+   6800 recv                80  FF   7
+   6801 recvMC              80  FF   7
+   6802 recvUC              80  FF   6
+   6803 minimal_sched_f     80  FF   7
+```
+
+### minimal_sched_fifo_spin_thread
+
+For this example, we will be configuring the scheduling policy of a thread used to spin the 
+executor. In this case, the thread middleware threads will not inherit these settings because
+the settings apply only to the spinner thread.
+
+```bash
+$ ros2 run minimal_scheduling minimal_sched_fifo_spin_thread
+```
+
+```bash
+$ ps -C minimal_sched_f -L -o tid,comm,rtprio,cls,psr
+    TID COMMAND         RTPRIO CLS PSR
+   6822 minimal_sched_f      -  TS   1
+   6823 minimal_sched_f      -  TS   7
+   6824 gc                   -  TS   5
+   6825 dq.builtins          -  TS   7
+   6826 dq.user              -  TS   2
+   6827 tev                  -  TS   0
+   6828 recv                 -  TS   2
+   6829 recvMC               -  TS   0
+   6830 recvUC               -  TS   7
+   6831 minimal_sched_f      -  TS   0
+   6832 minimal_sched_f     80  FF   7
+```
+
+Use `htop` or `ps` to observe the priorities of the threads.
+
+### minimal_sched_fifo_callback_group
+
+For this example, we will be using multiple callback groups to separate callbacks, so it is
+possible to spin them in threads with different scheduling priorities.
+
+```bash
+$ ros2 run minimal_scheduling minimal_sched_fifo_callback_group
+```
+
+Use `htop` or `ps` to observe the priorities of the threads.
+
+```bash
+$ ps -C minimal_sched_f -L -o tid,comm,rtprio,cls,psr
+    TID COMMAND         RTPRIO CLS PSR
+   6847 minimal_sched_f      -  TS   1
+   6848 minimal_sched_f      -  TS   0
+   6849 gc                   -  TS   1
+   6850 dq.builtins          -  TS   1
+   6851 dq.user              -  TS   7
+   6852 tev                  -  TS   1
+   6853 recv                 -  TS   3
+   6854 recvMC               -  TS   0
+   6855 recvUC               -  TS   7
+   6856 minimal_sched_f      -  TS   0
+   6857 minimal_sched_f     80  FF   0
+```
+
```

Comment:
The idea was to keep it simple and just show the configuration. Then we could create a more complete demo and show the effects. However, we did something that with the memory lock example. If it's simple enough maybe it is worth including some latency or missed deadlines measurements.

Reply:
@ralph-lange @shuhaowu I think I found a simple way to show the effect with and without configuring the priority. We can simply count the number of involuntary context switches (https://linux.die.net/man/2/getrusage). With stress and `SCHED_OTHER` the count should increase over time. With `SCHED_FIFO` the count should be zero or increase slowly. 

I ran a couple of experiments and this seems to be the case. I will update the example and check of the context switches counting is consistent in different platforms.

Reply:
Oh that's an interesting setting. I'll also try to add this into the metrics collection setup of my RT app framework and test it with you. Thanks for letting me know!

Reply:
I fully agree that the example should be kept as simple as possible. My idea with `stress-ng` was not to extend the code, but simply add a note at the bottom of the README saying something like:

_In addition to the changes to thread priorities verifiable by `ps`, you can also observe how the `Hello World` output of the unprioritized threads (and thus callbacks) slows down when you create a heavy computational load on your system using a stress tool like stress-ng (e.g., `stress-ng [our-proposal-for-a-configuration]`)._

Of course, such a note should only be added if a suitable configuration can be found where such effect is well observable.

Reply:
After some changes the example looks similar to the `minimal_memory_lock` one, we should the next involuntary context switches and we expect none if real-time settings are used. 

In the case of the callback group examples, it's possible to observe both non-real-time and real-time scheduling outputs at the same time. I recorded a preliminary asciinema so you get the idea. 

https://asciinema.org/a/wRvdwyndL5nzXOA2LmueZuOLf 

Please take a look at the latest implementation in case you have more input. Once it's good to go I will record more asciinema showing all the examples and the ps output in a side panel. 

Thanks for the valuable feedback btw.

Reply:
Extending the duration of the callback by re-using the [`burn_cpu_cycles`](https://github.com/ros2/examples/blob/master/rclcpp/executors/cbg_executor/src/examples_rclcpp_cbg_executor/pong_node.cpp#L75-L88)  makes the example more consistent and we don't need to stress the system to show the effects.

https://asciinema.org/a/WR31JDj6wKXKlhdjib4ffSah5